### DISCUSSION TITLE
Read image from GPU on background thread

### DISCUSSION THREAD

**Discussion by cbrl** [openscenegraph/OpenSceneGraph#1254] [Q&A] [answered]
Is it possible to download the rendered image from a secondary camera in a separate thread?

To give more details, I am creating a several cameras which render a single frame of the scene, then are removed from the scene graph. The rendered images are then read from the color attachment to be used elsewhere. I would like to read those images without stalling the main thread. Simply adding a final draw callback which reads the data from the color attachment causes hiccups.

This may be an OpenGL question in part, but if it's possible to do I'm not sure how it would work into the OSG API either.

**Comment by robertosfield** [comment, marked as answer]
OpenGL has a single threaded model where each graphics context can only deal with a single thread using it's FIFO at any one time.  The FIFO is processed by the driver in-order and calls like glReadPixels requires that all commands+data in  FIFO before it are fully processed before it can run, then finally there is the transfer from the GPU to main memory which is why you are seeing a hiccup and unfortunately why you can fix it.

If you can try to keep the images on GPU memory and process the images there.

Or adopt Vulkan/VulkanSceneGraph as it has a multi-threaded model.

**Reply by cbrl** [reply]
Unfortunately I can't leave the image on the GPU. I've heard of using a second shared OpenGL context to aid in multithreaded loading. Would that be applicable here?

I would like to eventually switch to VulkanSceneGraph. This type of thing would definitely be more ergonomic with Vulkan. However, my application is using osgEarth, so I would need to wait for additional development on Rocky before I can replace it.

**Reply by robertosfield** [reply]
You could try playing games with a shared graphics context and then try to run it multi-threaded by you'd need to use GL extensions to handle the synchronization.  OpenGL wasn't designed with this in mind, nor was the OSG, so you're on your own if you want to try this, I can't help you.

**Reply by cbrl** [reply]
Alright, I'll see if it's possible to work around this. Thanks for the quick response.
